### ISSUE TITLE
Models > Probability Distributions - third and final dialogue

### ISSUE THREAD

**Issue by rdstern** [africanmathsinitiative/R-Instat#1935]
This is a bit of a challenge for Lily and John, though it starts innocently enough.

The first step is to sort out the menu

Could Model > Probability Distributions > "Tables Plus" change to "Show Model".
And the second entry - now "Show Models" becomes "Compare Models".
The third entry stays as it was.

(Notice singular **Model** in first, while plural **Models** in the second menu item.)

The challenge is to undertake the third and last dialogue.  It is to construct the dialogue and the corresponding R code for the second entry, namely **"Compare Models"**.

As with the other dialogues in this set, it will use the different distributions that Lily has been working on, and where David and Danny may be editing the underlying R code.

Perhaps you can work on the dialogue and the R code in parallel?  Here are some points on the dialogue:
1.  It could be used with a single "distribution", i.e. it is also to show a distribution as well as to compare them.  (Some people may prefer it, because it uses a different R command to the "Tables Plus", now "Show Model".
2.  It will often be used with one "distribution type".  By that I mean a single distribution where one or more parameters are changed. For example t distribution with 4, 10 and 30 degrees of freedom, or normal distribution with mean 0, 10, 20 and s.d. 1, 2, i.e. 6 possibilities!
3. You begin to see the fun!  With more than one plot they could all be together or could use facets. Or a bit of each!  The challenge will partly be to make the dialogue easy to use.
4.  There could also be multiple distributions.  For example we may wish to compare the Poisson and Normal with the same mean, or normal and t, or gamma and log-normal.  I would be happy to start with just 2 distributions for this possibility.  I wonder then about putting the second distribution on a sub-dialogue.  That would then be easy to extend.  And the main dialogue could then be roughly the same as the current one for Tables Plus.  The new facility would be to be able to enter multiple numbers in each of the text boxes for the parameters.  So we could start with that.

John, perhaps you could start on the dialogue?  We could start without a sub-dialogue for now.

The issue on the R part is that there are (at least) two possible approaches.  Me may even want to use both, but I hope not.  They are as follows:

a) We continue with the Mosaic package, which is used for the current Tables Plus dialogue.  There is a command called plotdist in Mosaic which seems to do all we want. Yippee.  Except it doesn't use ggplot2. Boo! It uses lattice plots instead, so we could (possibly) still do all we want.  But do we want to take this route?
b) I would prefer to use the ggdistribution command in ggfortify.  This is easy to give in command mode, but you have to supply the x values, (which the Mosaic plotdist command would do for us).  And plotdist also detects automatically whether the distribution is discrete or continuous.  That is a more minor point perhaps.

I would prefer (of course) to use ggdistribution, but it would be good to explore both.  Lily, perhaps you could attack this aspect?  In ggdistribution here is a possible first try at working out the x-values.
1.  For each distribution on the main dialogue you calculate (behind the scenes) the probability of 0.001 and 0.999 and then take the smallest of the x values for 0.001 and the largest for 0.999.
2. That gives you the range, which you extend a bit at both ends.  So if the range is 0.5 to 20.5 you could extend by (say) 10%, or a total of 2 for this example - so 1 at each end.  So it is now -0.5 to 21.5.
3. What I don't know is what happens to the subsequent calculations if you look for a density or probability out of range.  I think it is OK, e.g. pexp(-1) = 0.  The zero boundary will be common, so perhap[s when the minimum calculated above is > 0, but the extended range is < 0 then make it 0.
4.  Then take perhaps 100 values within this range.

If you want to see some websites, then 
https://cran.r-project.org/web/packages/ggfortify/vignettes/plot_dist.html is for ggdistribution, while

https://cran.r-project.org/web/packages/mosaic/vignettes/GraphicsWithMosaic.html for Mosaic

Happy exploring!  

Roger

**Comment by lilyclements** [comment]
@Lunalo - I believe we agreed around a week ago for any changes to the main form to go through you and Steve to avoid conflicts. So I would suggest it is easiest for you to change those names on the main form! Obviously otherwise I am more than happy to - as long as you and Steve give me the go-ahead for changing the main form.

I will have a look at ggdistribution/ggfortify vs mosaic approaches tomorrow.

**Comment by Lunalo** [comment]
@lilyclements , I will do it immediately Steve's changes are merged. Thanks

**Comment by lilyclements** [comment]
So, I think we have settled that while ggdistribution is lovely, plotDist is the best course of action - at least for now. Perhaps in the future we can offer radio buttons to switch between ggdistributions and plotDist, but for now we will just use plotDist.

The dialog can be as follows

1. We can have the ucrDistwithParameters 
2. Below this we can have an option to graph any selection as 'Single' or 'Combine'. If 'Combine' is selected, then a label and a nud can come up where the user can choose how many columns they wish to have.
3. Below this we can have an option to "Add New Distribution" which will lead to a sub-dialog.
4. On the right hand side, we can give the option with check boxes of whether they wish to plot a Density, CDF or Histogram. Density is the default. 
5. Below this we can give a check box to "change y limit", this is default unchecked. If they wish to change limits then a ylim option can come up. This can bring up two nud's one for the minimum y value and one for the maximum y value.
@rdstern , shall we have one for x as well? I do not know if this would get too complicated with default values.
6. Then the usual ucrBase


Now, as for the R-Code

The overall code being run is
plotDist("distributionname", param1= , param2 =  , ...)
for example

plotDist("norm", sd=1, mean=0)

For part 4, a parameter will be added of 
kind = "density"
kind = "cdf"
or kind = "histogram"
Something about this part here is that if the user selects more than one checkbox, then the code has to run separately. That is, say the user selects Density and CDF for a normal distribution, we have to run
plotDist("norm", kind = "density")
plotDist("norm", kind = "cdf")
As two separate graphs

For part 5, another parameter is added. This is
ylim=c(MINVALUE, MAXVALUE)


Part 2 looks at combining the graphs. If the user chooses to combine the graphs then we are running a new function. This is
grid.arrange(x = plotDist(...), x = plotDist(...), etc)
where the plotDist(...) is the R-Code generated in part 1,4,5.

If the user selects to change number of columns then the parameter
ncol=*value in nud* is added.



A final feature, which Roger touched on at the start is an alteration to the ucrDistwithParameters is that it would be nice to have in that you can put in several values for parameters. For example, for the standard deviation you can ask to have in the standard deviation as 1, 3, 5. This can then yield three different graphs with the sd once as 1, then as 3 and then as 5. This can enable the user to see how a change in sd affects the distribution.


If this all makes sense then I will explain through the sub dialog! Let me know if you have any questions!

**Comment by rdstern** [comment]
Danny has now changed the menu items as suggested at the start of this issue.

**Comment by Lunalo** [comment]
@dannyparsons 

I am not sure if you are done with editing the user control for distribution. I have  noticed, unlike in mosaic which produces some nice colours automatically, ggdistribution produces plots with no fill and we can add that as one of the parameters in the ucrdistribution

**Comment by Lunalo** [comment]
@lilyclements 

I agree that plotdist looks better. I was just wondering if we want some consistency in creation of these plots. I think ggdistribution is better approach since we have implemented  graphics using ggplot2.

Thanks

**Comment by lilyclements** [comment]
@Lunalo 
I do agree that the consistency would be ideal, and I think it would be good if we could offer both ggdist and plotDist in the future, however, as a teaching tool I feel plotDist helps to differentiate between discrete and continuous distributions a lot nicer

**Comment by rdstern** [comment]
@Lunalo and @lilyclements

I like this discussion.  I too would like both approaches in the future.  But we have so much to do now that I am happy if the simpler approach is working now.  And I am happy that we continue to use the Mosaic commands, which will help when we write to them.

What is the current state of this dialogue?  Can it be enabled yet?  I would like to see it "in action"

**Comment by rdstern** [comment]
John, can you update us on the state of this dialogue.  It would "complete" the set of initial dialogues in the Model menu.

It would be good if there were something for the beta release.  If it isn't working, but the dialogue is in place, then it could be included now and Lily can add the label about the OK not being enabled.

**Comment by lilyclements** [comment]
Okay, I will add the OK not being enabled button once John has finished the work on it for the beta release to avoid a conflict. John, can you let me know when you've merged it so I can add this button?

**Comment by dannyparsons** [comment]
@lilyclements is this useful for model discussions? Otherwise I can close it.

**Comment by lilyclements** [comment]
Yes this would be good to discuss

**Comment by dannyparsons** [comment]
@lilyclements @stevenndungu anything left on this?

**Comment by stevenndungu** [comment]
Looking on show model pull request

**Comment by dannyparsons** [comment]
This was merged so will close this as specific issues can be opened in new issues
